### PR TITLE
crypto/pem/pem_lib.c: Add check for BIO_read

### PR DESCRIPTION
**Pull request by JiangJias**
As the potential failure of the BIO_read(),
it should be better to add the check and return
error if fails.
Also, in order to decrease the same code, using
'out_free' will be better.

Signed-off-by: Jiasheng Jiang <jiasheng@iscas.ac.cn>

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Review by paulidale — APPROVED** [review]

File: crypto/pem/pem_lib.c [hunk 1]
```diff
@@ -971,19 +971,21 @@ int PEM_read_bio_ex(BIO *bp, char **name_out, char **header,
...
+    if (BIO_read(dataB, *data, len) != len)
+        goto out_free;
     *len_out = len;
     *name_out = name;
     name = NULL;
     ret = 1;
 
+out_free:
+    pem_free(*header, flags, 0);
+    pem_free(*data, flags, 0);
```
Comment (t8m) [inline crypto/pem/pem_lib.c:988]: These cannot be called in case of success! The CI failure is relevant.

**Review by t8m — CHANGES_REQUESTED** [review]

**Comment by t8m** [comment]
This is still failing CI. You'll have to investigate why.

**Comment by JiangJias** [comment]
> 

Fine, I will solve it.

File: crypto/pem/pem_lib.c (hunk 1 above)
Comment (t8m) [reply crypto/pem/pem_lib.c:989]: Perhaps it could help to set *header and *data to NULL?

Comment (JiangJias) [reply crypto/pem/pem_lib.c:989]: Failed again.
I will try to find out why.

**Comment by JiangJias** [comment]
> This is still failing CI. You'll have to investigate why.

I have found out that the "headerlen" is always zero when tests fail.
So probably the reason is the wrong length returned by BIO_get_mem_data().
Can you give me some advice to help me solve the problem more quickly?

File: crypto/pem/pem_lib.c [hunk 2]
```diff
@@ -971,19 +971,25 @@ int PEM_read_bio_ex(BIO *bp, char **name_out, char **header,
     headerlen = BIO_get_mem_data(headerB, NULL);
     *header = pem_malloc(headerlen + 1, flags);
     *data = pem_malloc(len, flags);
-    if (*header == NULL || *data == NULL) {
-        pem_free(*header, flags, 0);
-        pem_free(*data, flags, 0);
-        goto end;
+    if (*header == NULL || *data == NULL)
+        goto out_free;
+    if (BIO_read(headerB, *header, headerlen) != headerlen) {
```
Comment (t8m) [inline crypto/pem/pem_lib.c:976]: Then just replace this condition with `headerlen != 0 && BIO_read(headerB, *header, headerlen) != headerlen`

Comment (t8m) [reply crypto/pem/pem_lib.c:976]: There is no point in calling the BIO_read for 0 bytes.

File: crypto/pem/pem_lib.c [hunk 3]
```diff
@@ -971,19 +971,22 @@ int PEM_read_bio_ex(BIO *bp, char **name_out, char **header,
...
     *data = pem_malloc(len, flags);
-    if (*header == NULL || *data == NULL) {
-        pem_free(*header, flags, 0);
-        pem_free(*data, flags, 0);
-        goto end;
-    }
-    BIO_read(headerB, *header, headerlen);
+    if (*header == NULL || *data == NULL)
+        goto out_free;
+    if (headerlen == 0 || BIO_read(headerB, *header, headerlen) != headerlen)
```
Comment (t8m) [inline crypto/pem/pem_lib.c:976]: This should have `&&` instead of `||`

Comment (JiangJias) [reply crypto/pem/pem_lib.c]: Thanks, it does work.

**Review by t8m — APPROVED** [review]

**Review by paulidale — APPROVED** [review]

**Comment by t8m** [comment]
Merged to master and 3.0 branches. Thank you.

**Pull request closed without merging**